_posts/2023-02-13-semantic-search-solutions.md

```diff
@@ -0,0 +1,106 @@
+
+### Option 1: Semantic search with a pretrained model
+
+A readily available pretrained model requires minimal setup and saves you time and effort of training. Use the following steps to build a pretrained solution in OpenSearch:
+
+1. **Choose a model**. We recommend the [TAS-B](https://huggingface.co/sentence-transformers/msmarco-distilbert-base-tas-b) model that is publicly available on HuggingFace. This model maps every document to a 768-dimensional vector. It has been trained on the [MS Marco](https://huggingface.co/datasets/ms_marco) dataset and shows impressive performance on datasets that belong to domains outside of MS Marco [[BEIR 2021](https://arxiv.org/abs/2104.08663)]. 
+2. **Make sure the model is in a format suitable for high performance environments**. You can [download the TAS-B model and some other popular models](https://github.com/opensearch-project/ml-commons/tree/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding) that are already in the correct format. If you are using another model, download the model and then follow the instructions in the [Demo Notebook to trace Sentence Transformers model](https://opensearch-project.github.io/opensearch-py-ml/examples/demo_tracing_model_torchscript_onnx.html) to convert it to a format suitable for high performance environments, such as [TorchScript](https://pytorch.org/docs/stable/jit.html) or [ONNX](https://onnx.ai/).
+3. **Upload the model to an OpenSearch cluster**. Use the [model-serving framework](https://opensearch.org/docs/latest/ml-commons-plugin/model-serving-framework/) to upload the model to an OpenSearch cluster, where it will create a vector index for the documents in the dataset. 
+4. **Search using this model**. Create a k-NN search at query time with the OpenSearch [Neural Search plugin](https://opensearch.org/docs/2.5/neural-search-plugin/index/). For additional details, follow the steps in the [Similar document search with OpenSearch](https://opensearch.org/blog/similar-document-search/#:~:text=%5D%0A%20%20%7D%0A%7D-,Search%20with%20k%2DNN,-In%20this%20example) blog.
+
+If you’re interested in semantic search performance, look for detailed benchmarking documentation that we’ll be releasing. This documentation will include such benchmarking data as query latency, ingestion latency, and query throughput.
+
+### Option 2: Semantic search with a tuned model
+
+To build a custom solution, ideally, you need a dataset that consists of `(query, relevant passage)` pairs from the chosen domain to train a model so that it performs well on that domain. In the OpenSearch context, it is common to have passages, but not queries. The synthetic query generation technique circumvents this problem by automatically generating artificial queries. For example, given the passage on the left, the synthetic query generator automatically generates a question similar to the one shown on the right.
```

Comment:
```suggestion
To build a custom solution, you ideally should have a dataset that consists of `(query, relevant passage)` pairs from the chosen domain in order to train a model so that it performs well on that domain. In the context of OpenSearch, it is common to have passages but not queries. The synthetic query generation technique circumvents this problem by automatically generating artificial queries. For example, given the passage on the left in the following image, the synthetic query generator automatically generates a question similar to the one shown on the right.
```